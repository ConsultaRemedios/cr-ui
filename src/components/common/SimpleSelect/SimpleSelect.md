SimpleSelect is a simple select component. We use the native options instead of simulating it
through a fake dropdown. Compared to BaseSelect, it is not so flexible, but it provides a better
accessibility, user experience and it is simpler to use.

```js
let data = { selected: 'f' };

const options = [
  { label: 'Feminino', value: 'f' },
  { label: 'Masculino', value: 'm' }
];

const onChange = (e) => {
  data.selected = e.value;
};

<div>
  <strong>Selected value:</strong> {{ data.selected }}

  <br/>
  <br/>

  <SimpleSelect
    name="gender"
    :options="options"
    :selected="data.selected"
    @change="onChange"
  />
</div>
```
