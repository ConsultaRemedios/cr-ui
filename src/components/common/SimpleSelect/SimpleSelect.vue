<template>
  <div :class="$style.wrapper">
    <div :class="$style.select">
      <span :class="$style.selectText">
        <span>{{ label }}</span>
      </span>

      <span :class="$style.selectIcon">
        <BaseIcon :id="icon.id" />
      </span>
    </div>

    <select
      :class="$style.nativeSelect"
      :name="name"
      v-bind="$attrs"
      @change="onChange"
    >
      <option
        disabled
        :selected="!value"
      >
        {{ placeholder }}
      </option>

      <option
        v-for="(option, index) in options"
        :key="index"
        :value="option.value"
        :selected="value === option.value"
      >
        {{ option.label }}
      </option>
    </select>
  </div>
</template>

<script>
import BaseIcon from '../BaseIcon';
import arrowIcon from '../../../icons/drop.icon.svg';

export default {
  name: 'SimpleSelect',
  components: { BaseIcon },
  inheritAttrs: false,

  props: {
    name: { type: String },
    model: { type: String },

    options: {
      type: Array,
      default: () => ([]),
    },

    selected: {
      type: String,
      default: '',
    },

    placeholder: {
      type: String,
      default: 'Selecione',
    },
  },

  data() {
    return {
      value: this.selected,
    };
  },

  computed: {
    label() {
      const option = this.options.find((o) => o.value === this.value);
      return option ? option.label : this.placeholder;
    },

    icon() {
      return arrowIcon;
    },
  },

  methods: {
    onChange({ target }) {
      const { value } = target;

      this.value = value;

      this.$emit('change', {
        name: this.model || this.name || '',
        value,
      });
    },
  },
};
</script>

<style module>
  .wrapper {
    position: relative;
  }

  .select {
    display: flex;
    align-items: center;
    height: 45px;
    padding: 0 15px;
    background: #FFF;
    border: 1px solid #DADADA;
    border-radius: 3px;
  }

  .selectText {
    font-size: 15px;
    color: #333333;
    flex-grow: 1;
    max-width: calc(100% - 24px);
  }

  .selectText > span {
    display: block;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .nativeSelect {
    cursor: pointer;
    appearance: none;
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    width: 100%;
    height: 100%;
    padding: 10px;
    opacity: 0;
  }
</style>
