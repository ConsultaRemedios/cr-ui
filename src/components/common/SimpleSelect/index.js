import SimpleSelect from './SimpleSelect.vue';

export default SimpleSelect;
