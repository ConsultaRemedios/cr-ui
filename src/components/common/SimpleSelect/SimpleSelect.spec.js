import { mount } from '@vue/test-utils';
import SimpleSelect from './SimpleSelect.vue';

const getSelected = wrapper => wrapper
  .findAll('option').wrappers
  .find(w => w.element.selected)
  .element;

describe('SimpleSelect component', () => {
  let propsData;

  beforeEach(() => {
    propsData = {
      name: 'gender',
      selected: 'F',
      placeholder: 'Selecione',
      options: [
        { label: 'Feminino', value: 'F' },
        { label: 'Masculino', value: 'M' },
      ],
    };
  });

  describe('Props', () => {
    describe('#options', () => {
      it('returns an empty array when the prop is not passed to it', () => {
        const wrapper = mount(SimpleSelect);
        expect(wrapper.vm.options).toEqual([]);
      });
    });

    describe('#selected', () => {
      it('returns an empty string when the prop is not passed to it', () => {
        const wrapper = mount(SimpleSelect);
        expect(wrapper.vm.selected).toEqual('');
      });

      it('changes selected option when prop "selected" is changed', async () => {
        const wrapper = mount(SimpleSelect, { propsData });
        const option = getSelected(wrapper);

        expect(option.value).toEqual('F');

        await wrapper.setProps({ selected: 'M' });

        expect(getSelected(wrapper).value).toEqual('M');
      })
    });

    describe('#placeholder', () => {
      it('returns "Selecione" when the prop is not passed to it', () => {
        const wrapper = mount(SimpleSelect);
        expect(wrapper.vm.placeholder).toEqual('Selecione');
      });
    });
  });

  it('matches the snapshot', () => {
    const wrapper = mount(SimpleSelect, { propsData });
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('Template contracts', () => {
    it('selects the placeholder option when selected prop is empty', () => {
      propsData.selected = '';

      const wrapper = mount(SimpleSelect, { propsData });
      const option = getSelected(wrapper);

      expect(option.value).toEqual('Selecione');
    });

    it('selects the proper option when selected prop is not empty', () => {
      const wrapper = mount(SimpleSelect, { propsData });
      const option = getSelected(wrapper);

      expect(option.value).toEqual('F');
    });

    it('selects the proper option when selected prop is changed', async () => {
      const wrapper = mount(SimpleSelect, { propsData });

      wrapper.find('option[value="M"]').trigger('change');
      await wrapper.vm.$nextTick();

      const option = getSelected(wrapper);

      expect(option.value).toEqual('M');
    });
  });

  describe('Event listeners', () => {
    it('emits #onChange when select is changed', () => {
      const wrapper = mount(SimpleSelect, { propsData });

      jest.spyOn(wrapper.vm, '$emit');

      expect(wrapper.vm.$emit).not.toHaveBeenCalled();

      wrapper.find('option[value="M"]').trigger('change');

      expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
      expect(wrapper.vm.$emit.mock.calls[0][1]).toEqual({name: "gender", value: "M"});
    });
  });

  describe('Initial data', () => {
    it('sets value data to equal the selected prop', () => {
      const wrapper = mount(SimpleSelect, { propsData });
      expect(wrapper.vm.value).toEqual('F');
    });
  });

  describe('Methods', () => {
    describe('#onChange', () => {
      const event = { target: { value: 'M' } };

      it('updates value', () => {
        const wrapper = mount(SimpleSelect, { propsData });

        wrapper.vm.onChange(event);

        expect(wrapper.vm.value).toEqual('M');
      });

      it('emits "change" event with proper params when it has model and name props', () => {
        propsData.model = 'gender-model';

        const wrapper = mount(SimpleSelect, { propsData });

        jest.spyOn(wrapper.vm, '$emit');

        wrapper.vm.onChange(event);

        expect(wrapper.vm.$emit).toHaveBeenCalledWith('change', {
          name: 'gender-model', value: 'M'
        });
      });

      it('emits "change" event with proper params when it has no model, but it has name prop', () => {
        const wrapper = mount(SimpleSelect, { propsData });

        jest.spyOn(wrapper.vm, '$emit');

        wrapper.vm.onChange(event);

        expect(wrapper.vm.$emit).toHaveBeenCalledWith('change', {
          name: 'gender', value: 'M'
        });
      });

      it('emits "change" event with proper params when it has no model and name props neither', () => {
        propsData.name = undefined;

        const wrapper = mount(SimpleSelect, { propsData });

        jest.spyOn(wrapper.vm, '$emit');

        wrapper.vm.onChange(event);

        expect(wrapper.vm.$emit).toHaveBeenCalledWith('change', {
          name: '', value: 'M'
        });
      });
    });
  });

  describe('Computed', () => {
    describe('#label', () => {
      it('returns the label of the selected option if there is one', () => {
        const wrapper = mount(SimpleSelect, { propsData });
        expect(wrapper.vm.label).toEqual('Feminino');
      });

      it('returns placeholder value if there is no selected option', () => {
        propsData.selected = undefined;

        const wrapper = mount(SimpleSelect, { propsData });
        expect(wrapper.vm.label).toEqual('Selecione');
      });
    });

    describe('#icon', () => {
      it('returns arrow icon', () => {
        const wrapper = mount(SimpleSelect, { propsData });

        expect(wrapper.vm.icon).toEqual({
          id: 'drop.icon.svg',
          viewbox: '0 0 24 24',
        });
      });
    });
  });
});
