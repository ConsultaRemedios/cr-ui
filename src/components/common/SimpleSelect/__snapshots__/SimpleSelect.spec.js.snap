// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleSelect component matches the snapshot 1`] = `
<div
  class="wrapper"
>
  <div
    class="select"
  >
    <span
      class="selectText"
    >
      <span>
        Feminino
      </span>
    </span>
     
    <span>
      <svg
        class="icon"
        viewBox="0 0 24 24"
      >
        <use
          xlink:href="#drop.icon.svg"
        />
      </svg>
    </span>
  </div>
   
  <select
    class="nativeSelect"
    name="gender"
  >
    <option
      disabled="disabled"
    >
      
      Selecione
    
    </option>
     
    <option
      value="F"
    >
      
      Feminino
    
    </option>
    <option
      value="M"
    >
      
      Masculino
    
    </option>
  </select>
</div>
`;
